### PR TITLE
feat: mypy for all type check

### LINKED ISSUE #10928: make type hint passed by mypy check for all the codebase in api/
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

for protect the bug

### 2. Additional context or comments

_No response_

### 3. Can you help us with this feature?

- [X] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by yihong0618**
# Summary

This maybe a huge Pull Request, the main goal is to make all the `mypy` type check passed to fix the potential bugs
So its still WIP and will be draft. and try to close https://github.com/langgenius/dify/issues/10928

the first steps had beed done. -> let `mypy` work in `api` folder.
And we fixed some of them,

you can check by

- gh pr checkout 10921
- when found  small bug fix in main branch first
- install mypy
- cd api && mypy .

- [x] make it work for mypy
- [x] add ci for it
- [x] always merge main for the latest code
- [x] fixed them folder by folder, maybe type by type

for now after this commit it still has 4261 errors

some details update here:

- https://github.com/pallets-eco/flask-sqlalchemy/issues/1327

image

@laipz8200 @crazywoola 

> [!Tip]
> Close issue syntax: `Fixes #<issue number>` or `Resolves #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.

# Screenshots

<table>
  <tr>
  <td>Before: </td>
  <td>After: </td>
  </tr>
  <tr>
  <td>...</td>
  <td>...</td>
  </tr>
</table>

# Checklist

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

### CODE REVIEW

**Comment by laipz8200** [comment]
It looks like a great piece of work! Just so you know, due to some historical reasons, the type annotations in the current codebase are not always accurate. Also, extra caution is needed when dealing with dict types from requests or the database.

If you need any assistance, feel free to reach out to me.

**Comment by laipz8200** [comment]
Additionally, according to our [Contribution Guide](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md), I recommend linking an issue to this PR to clarify your objective and avoid others submitting duplicate PRs.

**Comment by yihong0618** [comment]
> It looks like a great piece of work! Just so you know, due to some historical reasons, the type annotations in the current codebase are not always accurate. Also, extra caution is needed when dealing with dict types from requests or the database.
> 
> If you need any assistance, feel free to reach out to me.

yes, maybe you can also push commit to this too, for the fixing them all, and I would create an issue for it

**Comment by laipz8200** [comment]
For Flask-SQLAlchemy, I’d like to share our plan with you:

1. Replace the direct use of `db.session` with `Session` from `sqlalchemy.orm` for queries and updates.
2. In a future major release, we plan to have database models directly inherit from SQLAlchemy’s base class.
3. Finally, we aim to use SQLAlchemy directly instead of Flask-SQLAlchemy.

As a result, when encountering type-related issues, we can proceed with change 1 and temporarily ignore the errors caused by change 2.

**Comment by yihong0618** [comment]
> For Flask-SQLAlchemy, I’d like to share our plan with you:
> 
> 1. Replace the direct use of `db.session` with `Session` from `sqlalchemy.orm` for queries and updates.
> 2. In a future major release, we plan to have database models directly inherit from SQLAlchemy’s base class.
> 3. Finally, we aim to use SQLAlchemy directly instead of Flask-SQLAlchemy.
> 
> As a result, when encountering type-related issues, we can proceed with change 1 and temporarily ignore the errors caused by change 2.

yes, this time I also choose ignore it, thanks for the info

**Comment by bowenliang123** [comment]
With my -1 for this PR.

- mypy for static type check brings little benefits in project
- the in-line suppressions puzzles the developers, for what rule it's violating or how we could further fix it.
- no proper auto fixes tool to align with the mypy rules
- too many suppressions in this PR
- for the field types, suggest to use Pydantic which runs check in runtime

**Comment by yihong0618** [comment]
> With my -1 for this PR.
> 
> * mypy for static type check brings little benefits in project
> * the in-line suppressions puzzles the developers, for what rule it's violating or how we could further fix it.
> * no proper auto fixes tool to align with the mypy rules
> * too many suppressions in this PR
> * for the field types, suggest to use Pydantic which runs check in runtime

some are right, but for now, I checked many of the bugs caused by the type, if not fix it by now, the code base grows, it will never be fixed.

**Comment by yihong0618** [comment]
> > With my -1 for this PR.
> > 
> > * mypy for static type check brings little benefits in project
> > * the in-line suppressions puzzles the developers, for what rule it's violating or how we could further fix it.
> > * no proper auto fixes tool to align with the mypy rules
> > * too many suppressions in this PR
> > * for the field types, suggest to use Pydantic which runs check in runtime
> 
> some are right, but for now, I checked many of the bugs caused by the type, if not fix it by now, the code base grows, it will never be fixed.

but this part I totally agree maybe even all the code happy, it should not in the CI, its too still some pain for everyone work on the mypy, so keep it on develop side maybe better 
> the in-line suppressions puzzles the developers, for what rule it's violating or how we could further fix it.
no proper auto fixes tool to align with the mypy rules

**Comment by yihong0618** [comment]
Less than 1000 still working on it

**Comment by yihong0618** [comment]
less than 500 now

**Comment by yihong0618** [comment]
300 left

**Comment by yihong0618** [comment]
less than 200 now!

**Comment by laipz8200** [comment]
> less than 200 now!

Cool!

**Comment by yihong0618** [comment]
98 now

**Comment by laipz8200** [comment]
wow

**Comment by yihong0618** [comment]
Done 
This is phase 1
for phase 2 
- drop `Any` as many as possible
- add type for the types empty var
- fix all the FIXME for type error

**Comment by laipz8200** [comment]
Awesome! I will review it in these two weeks.

**Comment by yihong0618** [comment]
> Awesome! I will review it in these two weeks.

seems need to keep updating...

**Comment by laipz8200** [comment]
We also need to add the CI configuration in this PR.

File: .github/workflows/api-tests.yml [hunk 1]
```diff
@@ -56,6 +56,12 @@ jobs:
       - name: Run Tool
         run: poetry run -C api bash dev/pytest/pytest_tools.sh
 
+      - name: Run mypy
+        run: |
+          pushd api
+          poetry run -C api python -m mypy .
```
Comment (laipz8200) [inline .github/workflows/api-tests.yml:62]:
```suggestion
          poetry run python -m mypy .
```

File: .github/workflows/api-tests.yml [hunk 2]
```diff
@@ -56,6 +56,12 @@ jobs:
       - name: Run Tool
         run: poetry run -C api bash dev/pytest/pytest_tools.sh
 
+      - name: Run mypy
+        run: |
+          pushd api
```
Comment (laipz8200) [inline .github/workflows/api-tests.yml:61]: cool!

**Review by laipz8200 — APPROVED** [review]

**Pull request merged**